### PR TITLE
Feat/m floating i ps

### PR DESCRIPTION
**Pull request by Abhishekism9450**

### CODE REVIEW

File: nutanix/data_source_nutanix_floating_ips.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceNutanixFloatingIPsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (siddharth-kulshrestha) [inline nutanix/data_source_nutanix_floating_ips.go]:
```suggestion
func dataSourceNutanixFloatingIPsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```

File: nutanix/data_source_nutanix_floating_ips.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	v3 "github.com/terraform-providers/terraform-provider-nutanix/client/v3"
+)
+
+func dataSourceNutanixFloatingIPs() *schema.Resource {
```
Comment (siddharth-kulshrestha) [inline nutanix/data_source_nutanix_floating_ips.go:13]: By file name it looks like a plural datasource but from implementation it looks like a singular datasource.

File: nutanix/data_source_nutanix_floating_ips.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+	if err := d.Set("spec", flattenSpecFIP(resp.Spec)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(*resp.Metadata.UUID)
+	return nil
+}
+
+func flattenStatusFIP(stat *v3.FIPDefStatus) []interface{} {
```
Comment (siddharth-kulshrestha) [inline nutanix/data_source_nutanix_floating_ips.go]: Please rename this struct to FloatingIPDefStatus.

File: nutanix/provider.go [hunk 4]
```diff
@@ -162,6 +162,8 @@ func Provider() *schema.Provider {
 			"nutanix_foundation_central_imaged_node_details":  dataSourceFCImagedNodeDetails(),
 			"nutanix_vpc":                                     dataSourceNutanixVPC(),
 			"nutanix_vpc_list":                                dataSourceNutanixVPCList(),
+			"nutanix_floating_ips":                            dataSourceNutanixFloatingIPs(),
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go:168]: Both the datasource looks like plural.

File: nutanix/data_source_nutanix_floating_ips_list.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	v3 "github.com/terraform-providers/terraform-provider-nutanix/client/v3"
+	"github.com/terraform-providers/terraform-provider-nutanix/utils"
+)
+
+func dataSourceNutanixFloatingIPsList() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceNutanixFloatingIPsListRead,
+		Schema: map[string]*schema.Schema{
+			"length": {
```
Comment (siddharth-kulshrestha) [inline nutanix/data_source_nutanix_floating_ips_list.go]: Do we need offset and length as part of the schema ? 
I guess in other datasources we iteratively fetch all the entities, combine them and populate the .tf file. 

Keeping offset and length in terraform file doesn't make sense to me. If you have any specific reasons for doing it we can discuss.

File: nutanix/resource_nutanix_floating_ip.go [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+		}
+		return diag.Errorf("error while deleting Floating IP UUID(%s): %s", d.Id(), err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"DELETE_PENDING", "RUNNING"},
+		Target:     []string{"SUCCEEDED"},
+		Refresh:    taskStateRefreshFunc(conn, resp.Status.ExecutionContext.TaskUUID.(string)),
+		Timeout:    d.Timeout(schema.TimeoutDelete),
+		Delay:      FipDelayTime,
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_floating_ip.go:255]: Why do we have to use these global variables here ?

File: nutanix/resource_nutanix_floating_ip.go [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+		MinTimeout: FipMinTimeout,
+	}
+
+	if _, err := stateConf.WaitForStateContext(ctx); err != nil {
+		return diag.Errorf(
+			"error waiting for floating ip (%s) to delete: %s", d.Id(), err)
```
Comment (siddharth-kulshrestha) [inline nutanix/resource_nutanix_floating_ip.go]:
```suggestion
			"error waiting for floating ip (%s) to delete: %s", d.Id(), err.Error())
```

File: examples/floating_ip/main.tf [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+provider "nutanix"{
+  username = var.nutanix_username
+  password = var.nutanix_password
+  endpoint = var.nutanix_endpoint
+  port = 9440
+  insecure = true
+}
+
+
+# create Floating IP with External Subnet UUID
```
Comment (siddharth-kulshrestha) [inline examples/floating_ip/main.tf:20]: datasource example is missing.

File: nutanix/provider.go [hunk 9]
```diff
@@ -186,6 +188,7 @@ func Provider() *schema.Provider {
 			"nutanix_foundation_central_image_cluster": resourceNutanixFCImageCluster(),
 			"nutanix_foundation_central_api_keys":      resourceNutanixFCAPIKeys(),
 			"nutanix_vpc":                              resourceNutanixVPC(),
+			"nutanix_floating_ip":                      resourceNutanixFloatingIP(),
```
Comment (siddharth-kulshrestha) [inline nutanix/provider.go:194]: Documentation change is required.

**Review by siddharth-kulshrestha — CHANGES_REQUESTED** [review]

File: nutanix/data_source_nutanix_floating_ips_list.go (hunk 5 above)
Comment (Abhishekism9450) [reply nutanix/data_source_nutanix_floating_ips_list.go]: removed length and offset .

File: nutanix/resource_nutanix_floating_ip.go (hunk 6 above)
Comment (Abhishekism9450) [reply nutanix/resource_nutanix_floating_ip.go:255]: yes, some constant numbers are magic numbers which are pre checked in PR.

File: nutanix/provider.go (hunk 9 above)
Comment (Abhishekism9450) [reply nutanix/provider.go:194]: noted.

File: examples/floating_ip/main.tf (hunk 8 above)
Comment (Abhishekism9450) [reply examples/floating_ip/main.tf:20]: created datasource examples .

**Review by siddharth-kulshrestha — APPROVED** [review]

**Pull request merged**